api/v1alpha1/securitygroup_types.go

```diff
@@ -26,37 +26,85 @@ type RuleDirection string
+// +kubebuilder:validation:XValidation:rule="(!has(self.portRangeMin) || !has(self.portRangeMax))|| (self.portRangeMin <= self.portRangeMax)",message="PortRangeMax should be equal or greater than PortRangeMin"
+// +kubebuilder:validation:XValidation:rule="!(self.protocol == 'icmp' || self.protocol == 'icmpv6') || !has(self.portRangeMin)|| (self.portRangeMin >= 0 && self.portRangeMin <= 255)",message="When protocol is ICMP or ICMPv6 PortRangeMin should be between 0 and 255"
+// +kubebuilder:validation:XValidation:rule="!(self.protocol == 'icmp' || self.protocol == 'icmpv6') || !has(self.portRangeMax)|| (self.portRangeMax >= 0 && self.portRangeMax <= 255)",message="When protocol is ICMP or ICMPv6 PortRangeMax should be between 0 and 255"
+
+// +kubebuilder:validation:XValidation:rule="!has(self.remoteIPPrefix) || (isCIDR(self.remoteIPPrefix) && cidr(self.remoteIPPrefix).ip().family() == 4 && self.ethertype == 'IPv4') || (isCIDR(self.remoteIPPrefix) && cidr(self.remoteIPPrefix).ip().family() == 6 && self.ethertype == 'IPv6')",message="RemoteIPPrefix should be a valid CIDR and match the Ethertype"
 type SecurityGroupRule struct {
 	// Description of the existing resource
 	// +optional
+
 	Description *OpenStackDescription `json:"description,omitempty"`
 
 	// Direction represents the direction in which the security group rule
 	// is applied. Can be ingress or egress.
+	// +optional
 	Direction *RuleDirection `json:"direction,omitempty"`
 
-	// RemoteAddressGroupId (Not in gophercloud)
-
-	// RemoteIPPrefix
+	// RemoteIPPrefix is an IP address block. Should match the Ethertype (IPv4 or IPv6)
+	// +optional
 	RemoteIPPrefix *CIDR `json:"remoteIPPrefix,omitempty"`
 
-	// Protocol is the IP protocol can be represented by a string, an
-	// integer, or null
-	Protocol *Protocol `json:"protocol,omitempty"`
+	// Protocol is the IP protocol can be represented by a string or an
+	// integer represented as a string.
+	// +required
+	Protocol Protocol `json:"protocol,omitempty"`
 
-	// EtherType must be IPv4 or IPv6, and addresses represented in CIDR
+	// Ethertype must be IPv4 or IPv6, and addresses represented in CIDR
 	// must match the ingress or egress rules.
-	Ethertype *Ethertype `json:"ethertype,omitempty"`
-
-	PortRangeMin *int32 `json:"portRangeMin,omitempty"`
-	PortRangeMax *int32 `json:"portRangeMax,omitempty"`
+	// +kubebuilder:validation:Required
+	Ethertype *Ethertype `json:"ethertype"`
+	// If the protocol is [tcp, udp, dccp sctp,udplite] this value must be less than
+	// or equal to the PortRangeMax attribute value.
+	// If the protocol is ICMP, this value must be an ICMP code
+	// +optional
+	PortRangeMin *uint16 `json:"portRangeMin,omitempty"`
+	// If the protocol is [tcp, udp, dccp sctp,udplite] this value must be greater than
+	// or equal to the PortRangeMin attribute value.
+	// If the protocol is ICMP, this value must be an ICMP type
+	// +optional
+	PortRangeMax *uint16 `json:"portRangeMax,omitempty"`
```

Comment:
Don't use uints, they don't translate into an openapi schema in any way that correctly validates the unsigned nature, use `int32` and set a minimum value of zero using the appropriate kubebuilder marker